live_slam:
  pointCloudLeftTopic: ns2/velodyne_points
  pointCloudRightTopic: ns1/velodyne_points
  imuTopic: imu/data_raw
  odomTopic: odometry/imu
  encoderTopic: encoder_count
  altimeterTopic: altimeter_data
  gpsTopic: odometry/gps

  lidarFrame: "base_link"
  baseLinkFrame: "base_link"
  odometryFrame: "odom"
  mapFrame: "map"

  useImuHeadingInitialization: true
  useGpsElevation: false
  gpsCovThreshold: 2.0
  poseCovThreshold: 25.0

  savePCD: false
  savePCDDirectory: "/Documents/Project/CLion/live_slam_ws/src/live_slam/map"

  N_SCAN: 32
  Horizon_SCAN: 1800
  downSampleRate: 1
  lidarMinRange: 1.0
  lidarMaxRange: 100.0

  imuGyrBiasN: 4.848136811e-06
  imuGyrNoise: 1.745329252e-05
  imuAccBiasN: 1.46991e-05
  imuAccNoise: 5.87964e-05
  imuGravity: 9.7994
  imuRPYWeight: 0.05

  vehicleToLeftVelodyneTransform: [-0.516377, -0.702254, -0.490096, -0.334623,
                                    0.491997, -0.711704,  0.501414,  0.431973,
                                   -0.700923,  0.0177927, 0.713015,  1.94043,
                                    0.000000,  0.000000,  0.000000,  1.00000]

  vehicleToRightVelodyneTransform: [-0.514521,  0.701075, -0.493723, -0.333596,
                                    -0.492472, -0.712956, -0.499164, -0.373928,
                                    -0.701954, -0.0136853, 0.712091,  1.94377,
                                     0.000000,  0.000000,  0.000000,  1.00000]
  vehicleToImuTransform: [1.0, 0.0, 0.0, -0.07,
                          0.0, 1.0, 0.0,  0.0,
                          0.0, 0.0, 1.0,  1.7,
                          0.0, 0.0, 0.0,  1.0]
  vehicleToGpsTransform: [1.0, 0.0, 0.0, -0.32,
                          0.0, 1.0, 0.0,  0.0,
                          0.0, 0.0, 1.0,  1.7,
                          0.0, 0.0, 0.0,  1.0]

  edgeThreshold: 1.0
  surfThreshold: 0.05
  edgeFeatureMinValidNum: 10
  surfFeatureMinValidNum: 100

  # voxel filter paprams
  odometrySurfLeafSize: 0.4                     # default: 0.4 - outdoor, 0.2 - indoor
  mappingCornerLeafSize: 0.2                    # default: 0.2 - outdoor, 0.1 - indoor
  mappingSurfLeafSize: 0.4                      # default: 0.4 - outdoor, 0.2 - indoor

  # robot motion constraint (in case you are using a 2D robot)
  z_tollerance: 1000                            # meters
  rotation_tollerance: 1000                     # radians

  # CPU Params
  numberOfCores: 6                              # number of cores for mapping optimization
  mappingProcessInterval: 0.15                  # seconds, regulate mapping frequency

  # Surrounding map
  surroundingkeyframeAddingDistThreshold: 1.0   # meters, regulate keyframe adding threshold
  surroundingkeyframeAddingAngleThreshold: 0.2  # radians, regulate keyframe adding threshold
  surroundingKeyframeDensity: 2.0               # meters, downsample surrounding keyframe poses
  surroundingKeyframeSearchRadius: 50.0         # meters, within n meters scan-to-map optimization (when loop closure disabled)

  # Loop closure
  loopClosureEnableFlag: true
  loopClosureFrequency: 1.0                     # Hz, regulate loop closure constraint add frequency
  surroundingKeyframeSize: 50                   # submap size (when loop closure enabled)
  historyKeyframeSearchRadius: 25.0             # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 30.0           # seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 25                  # number of hostory key frames will be fused into a submap for loop closure
  historyKeyframeFitnessScore: 0.3              # icp threshold, the smaller the better alignment

  # Visualization
  globalMapVisualizationSearchRadius: 3000.0    # meters, global map visualization radius
  globalMapVisualizationPoseDensity: 10.0       # meters, global map visualization keyframe density
  globalMapVisualizationLeafSize: 1.0           # meters, global map visualization cloud density

  # Navsat (convert GPS coordinates to Cartesian)
#  live_slam_navsat:
#    frequency: 50
#    wait_for_datum: false
#    delay: 0.0
#    magnetic_declination_radians: 0
#    yaw_offset: 0
#    zero_altitude: true
#    broadcast_utm_transform: false
#    broadcast_utm_transform_as_parent_frame: false
#    publish_filtered_gps: false
#
#  # EKF for Navsat
#  live_slam_ekf_gps:
#    publish_tf: false
#    map_frame: map
#    odom_frame: odom
#    base_link_frame: base_link
#    world_frame: odom
#
#    frequency: 50
#    two_d_mode: false
#    sensor_timeout: 0.01
#    # -------------------------------------
#    # External IMU:
#    # -------------------------------------
#    # imu0: imu_correct
#    # make sure the input is aligned with ROS REP105. "imu_correct" is manually transformed by myself. EKF can also transform the data using tf between your imu and base_link
#    # imu0_config: [ false, false, false,
#    #               true,  true,  true,
#    #               false, false, false,
#    #               false, false, true,
#    #               true,  true,  true ]
#    #imu0_differential: false
#    #imu0_queue_size: 50
#    #imu0_remove_gravitational_acceleration: true
#    # -------------------------------------
#    # Odometry (From Navsat):
#    # -------------------------------------
#    odom0: gps/odom
#    odom0_config: [ true,  true,  true,
#                    false, false, false,
#                    false, false, false,
#                    false, false, false,
#                    false, false, false ]
#    odom0_differential: false
#    odom0_queue_size: 10
#
#    #                            x     y     z     r     p     y   x_dot  y_dot  z_dot  r_dot p_dot y_dot x_ddot y_ddot z_ddot
#    process_noise_covariance: [ 1.0,  0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
#                                0,    1.0,  0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
#                                0,    0,    10.0, 0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
#                                0,    0,    0,    0.03, 0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
#                                0,    0,    0,    0,    0.03, 0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
#                                0,    0,    0,    0,    0,    0.1,  0,     0,     0,     0,    0,    0,    0,    0,      0,
#                                0,    0,    0,    0,    0,    0,    0.25,  0,     0,     0,    0,    0,    0,    0,      0,
#                                0,    0,    0,    0,    0,    0,    0,     0.25,  0,     0,    0,    0,    0,    0,      0,
#                                0,    0,    0,    0,    0,    0,    0,     0,     0.04,  0,    0,    0,    0,    0,      0,
#                                0,    0,    0,    0,    0,    0,    0,     0,     0,     0.01, 0,    0,    0,    0,      0,
#                                0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0.01, 0,    0,    0,      0,
#                                0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0.5,  0,    0,      0,
#                                0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0.01, 0,      0,
#                                0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0.01,   0,
#                                0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0.015 ]
